### PR TITLE
Add msync support

### PR DESCRIPTION
**Pull request by phpnode**
First ever time writing C/C++, so apologies if I've made any terrible faux pas.

> Edit: This now works!

### CODE REVIEW

File: src/mmap.cc [hunk 1]
```diff
@@ -85,9 +86,55 @@ NAN_METHOD(AlignedAlloc) {
...
+    if (length <= offset) {
+      return NanThrowRangeError("Offset out of bounds");
+    }
+    else if (offset > 0 && (offset % getpagesize())) {
+      return NanThrowRangeError("Offset must be a multiple of page size");
+    }
+
+    data += offset;
+    length -= offset;
+
```
Comment (indutny) [inline src/mmap.cc]: Unnecessary newline.

File: src/mmap.cc [hunk 2]
```diff
@@ -85,9 +86,55 @@ NAN_METHOD(AlignedAlloc) {
...
+  char* data = node::Buffer::Data(args[0]->ToObject());
+  size_t length =  node::Buffer::Length(args[0]->ToObject());
+
+  if (args.Length() > 1) {
+    // optional argument: offset
+    const size_t offset = args[1]->Uint32Value();
+    if (length <= offset) {
+      return NanThrowRangeError("Offset out of bounds");
+    }
+    else if (offset > 0 && (offset % getpagesize())) {
```
Comment (indutny) [inline src/mmap.cc]: Please place `else` on a previous line here and below.

File: src/mmap.cc [hunk 3]
```diff
@@ -85,9 +86,55 @@ NAN_METHOD(AlignedAlloc) {
...
+  }
+
+  if (msync(data, length, flags) == 0) {
+    NanReturnValue(NanTrue());
+  }
+  else {
+    return NanThrowError(strerror(errno), errno);
+  }
+}
+
```
Comment (indutny) [inline src/mmap.cc]: Please add second newline here for consistency with code above.

File: src/mmap.cc [hunk 4]
```diff
@@ -85,9 +86,55 @@ NAN_METHOD(AlignedAlloc) {
...
+  int flags;
+  if (args.Length() > 3) {
+    // optional argument: flags
+    flags = args[3]->Uint32Value();
+  }
+  else {
+    flags = MS_SYNC;
+  }
+
+  if (msync(data, length, flags) == 0) {
```
Comment (indutny) [inline src/mmap.cc]: The `msync` man page says:

```
    The msync() system call writes modified whole pages back to the filesystem and updates the file modification time.  Only those pages containing addr and len-1 succeeding loca-
     tions will be examined.
```

If I get it right - to make this work properly you'll need to call `msync` on each page. You might want to use `getpagesize()` to figure out the size of the page, and do a loop, invoking `msync` on each iteration.

Comment (phpnode) [inline src/mmap.cc]: @indutny hmm, I read that definition differently - it says "writes modified whole pages" - not "writes a single page". Also in that case what is the `length` parameter for?

Comment (indutny) [inline src/mmap.cc]: @phpnode I'm actually talking about this phrase in particular:

```
Only those pages containing addr and len-1 succeeding loca-
     tions will be examined.
```

So other pages will be ignored, at least on the terms of the docs.

Comment (indutny) [inline src/mmap.cc]: Nah, it doesn't look like this in OS X code: https://github.com/opensource-apple/xnu/blob/10.10/osfmk/vm/vm_map.c#L13791-L14078

Comment (indutny) [inline src/mmap.cc]: Going to check FreeBSD and linux, just to make sure.

Comment (indutny) [inline src/mmap.cc]: Anyway, could you please rewrite the test in a way that it'll create a temporary file of size = `4 * page size` and test that one `msync()` will suffice?

Comment (phpnode) [inline src/mmap.cc]: @indutny I think the function api wouldn't make sense if those were the semantics, also from http://linux.die.net/man/3/msync

> The msync() function shall write all modified data to permanent storage locations, if any, in those whole pages containing any part of the address space of the process starting at address addr and continuing for len bytes.

Comment (indutny) [inline src/mmap.cc]: Cool, thanks for checking it out. So two things then:

* Test for multiple pages
* Style fixes

Thank you!

**Comment by phpnode** [comment]
@indutny PR updated, hope I got the style right

File: test/api-test.js [hunk 5]
```diff
@@ -53,4 +53,101 @@ describe('mmap.js', function() {
...
+      assert(mmap.sync(buf, mmap.PAGE_SIZE, mmap.PAGE_SIZE, mmap.MS_SYNC));
+      var content = fs.readFileSync(this.file, 'utf8');
+      assert(content.slice(0, 11) === 'Hello There');
+      assert(content.slice(-12) === 'Hello There ');
+      buf = null;
+      gc();
+    });
+
+    it('should sync 4 pages at once', function () {
+      var file = __dirname + '/test-4-pages.txt';
```
Comment (indutny) [inline test/api-test.js]: What do you think about using `os.tmpDir()` here?

Comment (indutny) [inline test/api-test.js]: Also, I suggest putting it into `beforeEach` and cleaning it up in `afterEach`.

File: test/api-test.js [hunk 6]
```diff
@@ -53,4 +53,101 @@ describe('mmap.js', function() {
...
+        fs.closeSync(fd);
+        assert(buf);
+      });
+
+      it('should reject invalid page size offsets', function () {
+        var threw = false;
+        try {
+          mmap.sync(buf, 2);
+        }
+        catch (e) {
```
Comment (indutny) [inline test/api-test.js]: Please put `catch` on a previous line, and below too.

File: src/mmap.cc [hunk 7]
```diff
@@ -85,9 +86,51 @@ NAN_METHOD(AlignedAlloc) {
...
+    NanReturnValue(NanTrue());
+  } else {
+    return NanThrowError(strerror(errno), errno);
+  }
+}
+
```
Comment (indutny) [inline src/mmap.cc]: Needs second newline ;)

**Comment by indutny** [comment]
Yay, we are close to it! Thanks!

File: test/api-test.js (hunk 5 above)
Comment (phpnode) [inline test/api-test.js]: will do, much nicer

File: src/mmap.cc (hunk 7 above)
Comment (phpnode) [inline src/mmap.cc]: @indutny i was confused about this one, where should the newline go?

Comment (indutny) [inline src/mmap.cc]: I'm using two newlines to separate functions definitions from each other in C++.

Comment (phpnode) [inline src/mmap.cc]: ah got it, thanks!

**Comment by indutny** [comment]
Landed and released, thank you!

**Comment by phpnode** [comment]
yay, thanks!

**Pull request closed without merging**